### ISSUE TITLE
Allow embedders to decorate ranges in the buffer

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#1852]
Creating this to centralize discussion on @mofux's idea to support monaco-like decorations as discussed in https://github.com/xtermjs/xterm.js/issues/1380#issuecomment-379848832 and https://github.com/xtermjs/xterm.js/issues/791#issuecomment-448372516.

**Comment by Tyriar** [comment]
Was just thinking, once https://github.com/xtermjs/xterm.js/issues/1176 is done if we had some view part that could be inserted then the 1337 iterm sequence to insert an image could be completely implemented within an addon.

**Comment by sdegutis** [comment]
@Tyriar I'm particularly interested in an xterm.js feature where plugins can get DOM elements added into the "flow" of the terminal grid and even between rows, which the plugin can then completely control by appending interactive elements into it. Thanks to your comment to the issue I just opened, I found and have been reading up on a lot of discussion between you, @mofux, and @PerBothner, and I see that xterm.js is very well positioned to be able to add this feature, but that it's not quite there yet. I may be able to allocate several days or perhaps even weeks of full time work towards helping to implement this feature within xterm.js, pending we can come up with a reasonable and concrete plan, and can come up with a way to coordinate effectively on it. Please let me know your thoughts. Thanks for your time :)

**Comment by Tyriar** [comment]
@sdegutis glad to hear you're interested in contributing! It's been a while since we talked about this but here are some thoughts:

- We probably don't want to  get into the business of inserting elements in between rows as that would involve a lot of changes in how things are rendered and also mess with the interactive viewport at the bottom of the buffer
- I'm not too familiar with the Monaco decoration API but it's tried and true and worth taking inspiration from. Note that unlike in Monaco, we don't necessarily have DOM elements for the cells (depends on rendererType)
- I'm not sure what the right behavior would be when rows/cell with decorations get resized and get wrapped, I guess the embedder is responsible for listening to `Terminal.onResize` and reacting/recreating for that as appropriate

Given this, my current thinking is that we could maybe have a couple of flavors of decorations:

- Arbitrary DOM elements whose positioning is handled by xterm.js, for example you could use an `<img>` to decorate some cell, when the user scrolls the viewport then xterm.js will position that (with the help of `Marker`s). This type would probably need an extra container with `overflow:hidden` such that a decoration could span multiple rows. 
- Changing the background color of a cell, this would need to be handed over to the renderer internally to override the regular background color so it's drawn underneath the text.

FYI my main use case for wanting this is to enable visual indicators for VS Code's command tracking feature where it flags rows as have run a command. This would include a static indicator (Terminal.app uses a small `[` icon in a gutter on the left of the cell) and flag a row by changing its background color and fading back briefly when navigating to commands.

**Comment by sdegutis** [comment]
@Tyriar Hmm I think what I'm looking for may be slightly different than this issue then. What you're describing sounds a lot more like Monaco's deltaDecorations API ([example](https://microsoft.github.io/monaco-editor/playground.html#interacting-with-the-editor-line-and-inline-decorations)), whereas what I want is closer to their addContentWidget example.

Specifically, what I'd like to do is add an empty \<div> (which the user will be able to add more elements to) over a rectangle of the grid, as if it were part of the character grid itself, so that it scrolls when the rest of the grid scrolls.

I didn't realize they meant two different things until I looked closer at Monaco's API and tried them both out in their playground. I'm still having some difficulties getting this working, but I can explain more in a new issue for widgets rather than this one for decorators, unless you think maybe there's enough implementation/feature overlap between the two that they could be combined.

**Comment by jerch** [comment]
@sdegutis I think you want something like the "grid-holes" idea [here](https://github.com/xtermjs/xterm.js/issues/1901#issuecomment-459448213). This is a bad idea in terms of compliance to the current terminal interface (mainly for cursor and viewport issues) and portability to other emulators in general. If you really need this, I fear it wont make into xterm.js itself due to these incompatibilities (thus only possible as your own xterm.js addon).

The second idea there ("make room in the grid") is conform to the terminal interface, maybe this would work for you. Ofc this cannot be extended after creation, thus no user appends possible or even freely resized. Prolly not want your looking for.

**Comment by sdegutis** [comment]
@jerch That's a similar idea but my plan involved making sure that the process "made room" for the widget, and currently widgets are full-width (e.g. `width: 100%` in CSS), so "making room" just means the process has to print "\n\n\n\n" for a 4-row-tall widget. There are strange questions like "what happens if the user moves the cursor onto a part of the grid where the widget is" but I think those can semantically be figured out...

**Comment by jerch** [comment]
@sdegutis Could you give a short explanation what you want to place there (content type and such)? With a terminal there are several things to consider, esp. security-wise.

Example:
If you want to place there HTML content retrieved from terminal side, the HTML actually gets loaded in the host machine context. This will earn you an CVE, if not clearly stated or cannot be disabled.

**Comment by sdegutis** [comment]
@jerch I had in mind built-in widgets that the terminal comes with, which the calling process could "activate", i.e. make it appear in the terminal's grid. Thus it's a sort of whitelist of widgets that already have been vetted for security.

I have this already working for the most part, but I'm having surprising amount of difficulty getting scrolling to work, due to how xterm.js does its scrolling and how I'm adding my widgets to xterm's own root `term.element`. That's mostly what I'm stuck on solving externally.

**Comment by jerch** [comment]
@sdegutis Ah ok, well security aside, this sounds doable by an additional output layer. I have similar plans for an image layer to get image output working someday. Maybe have a look here, I have no scrolling issues: https://github.com/xtermjs/xterm.js/blob/3c5c8be6369ef8a2c42508b7ae1f8259d7c99dc1/addons/xterm-addon-image/src/ImageStorage.ts#L288

**Comment by PerBothner** [comment]
(Slightly off-topic - but it gets relevant ...)

I've been thinking about how one could support variable-width fonts in a terminal.  For some (human) languages using a single-width (or even duo-width with wide characters) is awkward and/or considered very ugly.  Even for English it's considered ugly - note this "issue" page uses variable-width fonts!

It wouldn't take much to extend the traditional terminal model to support variable-width "character cells". Instead of a rectangular grid of cells, you have a list of lines, each consisting of a variable number of variable-width cells.  Special escape sequence could be used to move around in this mode.  I'm not going into details here, but there are some non-obvious advantages.  For example the application and terminal don't have to agree on which characters are double-width, as cursor positioning would be done by character (or glyph), rather than cell.  When terminal width is changed, the terminal (rather than the application) would handle re-break/re-wrapping. (The application cannot know the width the terminal will have when the re-display is done.)

Once you have variable-width characters, you might also allow variable-height characters, and thus also variable-height lines. (The height of a line is the height of the tallest character in a line.)

Back to the topic at hand: Once you have variable width and height of a character, you can use a "pseudo-character" that consists of an image.  Or you can combine a character with an annotation, including a custom "paint" rule.  So this gives you extreme flexibility for "decorating" a character.

Note this isn't terribly difficult to implement - and it need not break the traditional model. DomTerm mostly already supports variable-size characters and lines, while still keeping very solid xterm compatibility. The main thing missing is to specify escape sequences to make it easier for an application to benefit from this model.  I'm considering modifying fish shell to take advantage of such a mode (since I already did some work on fish redisplay).  (The main changes to fish would be to treat the terminal as infinite-width, delegating line-breaking to the terminal; no special handling for double-width characters; ignore sigwinch.)

**Comment by Tyriar** [comment]
> That's a similar idea but my plan involved making sure that the process "made room" for the widget, and currently widgets are full-width (e.g. width: 100% in CSS), so "making room" just means the process has to print "\n\n\n\n" for a 4-row-tall widget.

@sdegutis This is what I was thinking in my little write up; decorating using DOM nodes would be unrestricted. Then an outer container the size of the terminal would use `overflow:hidden` such that it can be rendered above and part of it would be cut off:

```
****** < div is drawn above viewport because it's taller than 1 row
******
---------- < top of viewport
******

---------- < bottom of viewport
```

**Comment by sdegutis** [comment]
@PerBothner that's an interesting approach. I'm not sure it's compatible with the other goals I have, but in general interesting. I think an easier middle ground for my purposes is to just "overlay" a widget on top of the grid of fixed-size-characters. But I really like the innovative thinking you have going on and have been following DomTerm with plenty of interest.

**Comment by sdegutis** [comment]
@Tyriar Oh ok I must have misunderstood at first. Yeah I think that would work well. I have something similar going on right now, that allows me to do stuff like this:

<img width="320" alt="Screen Shot 2019-11-29 at 2 11 22 PM" src="https://user-images.githubusercontent.com/4399718/69888628-826ece80-12b2-11ea-8ebd-36025de5c123.png">

The problem I have right now is related to scrolling. I'll check out how @jerch does it but the summary of the bug I'm seeing is that, when the terminal is large enough to scroll, then hovering over the widget and trying to scroll will get "hijacked" by the terminal which then scrolls instead. This is when adding the widget to xterm.js via `term.element.append(widget)`. I've been looking into how xterm.js implements scrolling but don't see anything unusual that would cause this, aside from the fact that `.xterm-scroll-area` is inside `xterm-viewport` which comes *before* `xterm-screen` in the DOM, so it confuses me how it gets scroll events instead of them being sent to `xterm-screen` or something inside it -- which points to me just really not understanding at all how xterm.js gets scrolling to work and how I can get my element to scroll with it.

**Comment by Tyriar** [comment]
@sdegutis if you provide the element can't you just add a higher z-index to make sure it's on top? The scroll stuff is a little weird as we're kind of hacking it a bit to get a native scroll bar on a virtualized container (the terminal buffer).

**Comment by sdegutis** [comment]
@Tyriar That doesn't work for some reason. I had an idea though this morning for an alternative scrolling technique to get native scrollbars that might help fix my problems: what if you just grow the \<canvas> along with the scrollable container, which it would then become a child of, and only draw the visible portion of it based on where the container is scrolled to? So we'd have something like:

```
|--scrollview-and-canvas--|
|                         |
|--viewport---------------|
| (only this is drawn)    |
|                         |
|                         |
|                         |
|-------------------------|
|                         |
|-------------------------|
```

**Comment by Tyriar** [comment]
@sdegutis you can't grow a canvas that big as I think you eventually (and quickly) hit the upper limit of the GPU texture size. If we did anything with scrolling I'd want to move it closer to a real custom scroll bar (integrating [ScrollableElement](https://github.com/microsoft/vscode/blob/master/src/vs/base/browser/ui/scrollbar/scrollableElement.ts) is possible). Also there's extra overhead, more layouts, more drawing, etc. that would happen when you scroll.

**Comment by sdegutis** [comment]
Ah good point.

**Comment by jerch** [comment]
offtopic:
@PerBothner This sounds interesting. I like the idea to think into this direction, the current model already carries way to much ifs and elses due to the fix width idea (esp. for wide chars and wrapping lines in general, both are cumbersome to deal with). Could this be made compatible with very low interfaces (like an early terminal driver during bootup directly sitting on video RAM/framebuffer)? Imho it should keep working with these low level interfaces as kinda the whole terminal interface is derived from that simplicity. Historically we had a differentiation between dumb terminals with very reduced capability set (mostly bound to the functionality of real ttys) and somewhat more intelligent ones (I'd count any VT100+ into the latter). Not sure how far the OS layers have evolved, I think all line disciplines are still stuck at the dumb level in ICANON mode and try to anticipate the width and char handling according to the width. This is somewhat questionable, but changing it is not easy as it will raise tons of compat questions (the ldisc is used for many driver interactions). Maybe extending the TTY spec will do, something like an advanced ICANON2 mode without width notion, were the terminal is free to handle wrapping stuff on its own. This would need several additional sequences (and prolly ioctls) to be properly announced to the terminal (driver/emulator) though. Also giving up control of exact output representation needs more thinking how to deal with that from app side.

**Comment by PerBothner** [comment]
(Off-topic with re-direction)

I wrote up an [initial sketch of my variable-width glyph idea](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/variable-sized-glyphs.md).

I also created an [issue on freedesktop.org](https://gitlab.freedesktop.org/terminal-wg/specifications/issues/21). That seems like a good place for further discussion and ideas relating to this topic.

**Comment by Tyriar** [comment]
I don't think this variable width idea is something that applies to this and am worried about the discussion derailing. I think we should just work with what we already have and overlay a decoration on top of a range of cells that the application has made room for.

**Comment by PerBothner** [comment]
I (mostly) agree - hence my previous message suggesting to move discussing [variable-width glyphs here](https://gitlab.freedesktop.org/terminal-wg/specifications/issues/21).

(FWIW: Long-term (after webgl is stable and the default) I could see re-implementing DomTerm on top of xterm.js. Doing so would be easier if variable-size characters and "glyphs" could be supported efficiently, perhaps as some kind of addon or extension, as longs as it's maintainable.)

**Comment by sdegutis** [comment]
Fwiw the ideal solution for my needs is something very similar to @jerch's second solution in [this comment](https://github.com/xtermjs/xterm.js/issues/1901#issuecomment-459448213). Right now I'm doing it externally but running into a ton of issues. That's part of why I opened #2588. Right now I don't have a good proposal for any solution, only a list of the external solutions I attempted, and their problems. Are there any other issues/comments besides the linked comment where you guys have discussed embedding foreign content (DOM element) in a sub-grid within the existing grid? That's really all I need.

**Comment by Tyriar** [comment]
@sdegutis you mean this one?

![image](https://user-images.githubusercontent.com/2193314/70259588-4ec4f600-1743-11ea-933e-565ae9f24b05.png)

That's essentially what I'm thinking as well, but it's up to the terminal program or manual calls to `Terminal.write` to make the room for the overlay. From xterm.js' perspective you just have some DOM element attached to the blue cell and that's most all it knows (it would also need to measure the DOM element so it can draw it when the blue cell is off screen).

**Comment by sdegutis** [comment]
@Tyriar Exactly that picture, yes. My program already does `term.write('\n'.repeat(numRows))` so that's not the issue for me. But I'm running into a surprising amount of problems overlaying a DOM element over `term.element`:

- The element draws over the scrollbars if it has a background color and is full-row.
- Scrolling the terminal will instantly halt when mouse is now above foreign node.
- Had to keep two overlay containers, one each for main/alt buffers (solved).
- Had to use [private API](https://github.com/xtermjs/xterm.js/blob/17cb610ef1eeeb96b35c2fea1d4f325d73797434/addons/xterm-addon-fit/src/FitAddon.ts#L59-L60) to [get cell width and height](https://github.com/xtermjs/xterm.js/blob/17cb610ef1eeeb96b35c2fea1d4f325d73797434/addons/xterm-addon-fit/src/FitAddon.ts#L77-L78) (solved but not ideal).
- Had to sync overlay containers' `marginTop` with the `top` of bounding rect of `.xterm-scroll-area`, which sometimes reports a number past the actual end of the element due to Apple's over-scroll feature, messing up foreign-overlay positions and causing strange behavior/bugs when trying to scroll terminal back to normal position. Currently working on this now but getting nowhere. If it's purely an Apple bug, then moving to [ScrollableElement](https://github.com/microsoft/vscode/blob/master/src/vs/base/browser/ui/scrollbar/scrollableElement.ts) might solve it completely.
- *...And more!* (Well, possibly.)

**Comment by Tyriar** [comment]
> The element draws over the scrollbars if it has a background color and is full-row.

I would expect all decorations to live under another element, say `.xterm-decorations` which is the size of the terminal (excluding the scroll bars) and uses `overflow:hidden`.

> Scrolling the terminal will instantly halt when mouse is now above foreign node.

Does this work?

```css
.xterm-decorations {
  pointer-events: none;
}
```

> Had to use private API to get cell width and height (solved but not ideal).

I want that to be public but the exact values to expose aren't clear yet: https://github.com/xtermjs/xterm.js/issues/702

> Had to sync overlay containers' marginTop with the top of bounding rect

Hmm, I don't think you would need to get into that if it was hosted in `.xterm-decorations`? Just position the elements absolutely based on `ydisp`?

**Comment by sdegutis** [comment]
> > The element draws over the scrollbars if it has a background color and is full-row.
> 
> I would expect all decorations to live under another element, say `.xterm-decorations` which is the size of the terminal (excluding the scroll bars) and uses `overflow:hidden`.

Yes that's what I currently have.

> > Scrolling the terminal will instantly halt when mouse is now above foreign node.
> 
> Does this work?
> 
> ```css
> .xterm-decorations {
>   pointer-events: none;
> }
> ```

Technically it fixes the "scrolling just stops" problem, but my goal is to be able to have interactive widgets within the DOM. On normal web pages, e.g. blog pages that have `overflow:auto` code snippets, I'm able to scroll past them, and the page doesn't stop scrolling just because the mouse happens to be over one of them for a split second while scrolling. This is the behavior that is somehow broken, possibly due to how xterm.js's scrolling is implemented.

> > Had to sync overlay containers' marginTop with the top of bounding rect
> 
> Hmm, I don't think you would need to get into that if it was hosted in `.xterm-decorations`? Just position the elements absolutely based on `ydisp`?

What I do right now is, I have a sibling \<div> that immediately follows `.xterm.terminal` in the DOM, and this is my "overlay container" (e.g. `.xterm-decorations`). This has absolutely positioned elements that don't move, and I only adjust the overlay container's `.marginTop` based on the scroll position of xterm.js's pseudo scroll element. This works great for positioning all the widgets correctly in most situations, but it's got the problems I mentioned above.

**Comment by Tyriar** [comment]
> This is the behavior that is somehow broken, possibly due to how xterm.js's scrolling is implemented.

Yeah I think I get the problem now, it has to be under `.xterm-viewport` in order for event bubbling to work I think. Have you tried moving `.xterm-decorations` that?

**Comment by sdegutis** [comment]
Will give it a shot tonight, thanks for the idea.

**Comment by sdegutis** [comment]
@Tyriar 

Demos:

- https://sdegutis.github.io/xterm-widget-test-case/idea1.html - demo of version with next-sibling to term.element

- https://sdegutis.github.io/xterm-widget-test-case/idea2.html - demo of version with child of term._core.viewport._viewportElement

Problems with the second one (your latest idea):

- It scrolls per-pixel along with xterm-viewport's scrollTop, which doesn't match the terminal itself which scrolls per-character. (Even after #1140 is done, presumably pixel-level scrolling would be an option, and widgets would still need to scroll per-character when it's disabled.)

- Scrolling *inside* the widget no longer works. No mouse events work, not selecting text or clicking links, nothing. The widget has pointer-events: none, z-index: 3.

- With z-index: 4, clicking links now does work, but selecting text and scrolling inside the widget still doesn't work.

**Comment by Tyriar** [comment]
Bringing over discussion from https://github.com/xtermjs/xterm.js/issues/2607:

We should attach the decorations to the buffer they were created in such that they hide when going into the alt buffer. I think alt buffer decorations should be more temporary and be disposed when the normal buffer is entered.

**Comment by Tyriar** [comment]
> It scrolls per-pixel along with xterm-viewport's scrollTop

Scrolling in Edge is very very slow because of a browser feature to base the scroll on the size of the container.

> Scrolling inside the widget no longer works.

Didn't you just say scrolling works? It would be good to know what OS/trackpad/browser you're using.

**Comment by sdegutis** [comment]
@Tyriar imagine if someone is inside vim and runs a vim plugin that adds a decoration, and does ctrl-z, types some stuff in the main terminal, then foregrounds vim again. This is admittedly a crazy edge case, so I'm not saying it should be taken super seriously. But if anything more plausible than this should be allowed, then you might not want to dispose of it on the alt buffer just yet.

Well, unless that goes against the nature of the alt buffer -- I remember reading in some thread somewhere today (could be [in here](https://gitlab.freedesktop.org/terminal-wg/specifications/issues/12)) that the alt buffer *always* gets disposed of when switching back to main buffer. So maybe vim just redraws everything when switching back to it? In which case, it'd just have to redraw its decoration... that is, assuming it doesn't have any state. Most decorations in Monaco don't, they're sort of pure functions, even the git diff ones like the links in `Accept Theirs | Accept Both | ...` are stateless. If this is *always* the case, then I don't see a problem with disposing of them.

**Comment by Tyriar** [comment]
> So maybe vim just redraws everything when switching back to it?

I _think_ this is what happens.

**Comment by sdegutis** [comment]
@Tyriar In those two demo files, I'm using both Chrome and Safari on macOS Catasomething.

**Comment by Tyriar** [comment]
@sdegutis macbook touchpad or one of those apple mice I'm assuming? They scroll like trackpads (by pixel) not by "line" (which chrome translates into 33px on Windows).

How scrolling works under the scenes is super confusing as it's left up to the browser and not specced out 😄

**Comment by sdegutis** [comment]
Oops yeah, forgot to mention, using a Macbook Pro trackpad.

But it should be the same even when you scroll with an old mouse as long as it can scroll by at least some arbitrary amount of pixels? (No mouse I know scrolls exactly the same number of pixels as a terminal cell height.)

Either way the concept is the same: when you scroll inside the widget (which intentionally has more content than its fixed height can show), nothing happens.

**Comment by sdegutis** [comment]
> > So maybe vim just redraws everything when switching back to it?
> 
> I _think_ this is what happens.

Just checked, yeah it does, never mind most of my tangent. But I guess the one thing to take away is that means decorations must be expected to be destroyed on buffer changes, at least in one direction, and so they should be encouraged to be stateless maybe?

**Comment by Tyriar** [comment]
> Either way the concept is the same: when you scroll inside the widget

Scrolls fine for me 🤷‍♂ 

> But it should be the same even when you scroll with an old mouse as long as it can scroll by at least some arbitrary amount of pixels?

On Windows a scroll "notch" scrolls 100px on Chrome which is used as an estimate to the default Windows setting of 3 lines, and a variable amount of Edge due that feature.

**Comment by sdegutis** [comment]
> > Either way the concept is the same: when you scroll inside the widget
> 
> Scrolls fine for me 🤷‍♂

Hmm. The red widget scrolls internally? In the second demo? In Chrome?

> > But it should be the same even when you scroll with an old mouse as long as it can scroll by at least some arbitrary amount of pixels?
> 
> On Windows a scroll "notch" scrolls 100px on Chrome which is used as an estimate to the default Windows setting of 3 lines, and a variable amount of Edge due that feature.

Oooh, so that probably explains why #1140 isn't as popular of a feature request as I'd thought it'd be -- people on Windows with regular mice don't really have the possibility of pixel scrolling.

**Comment by sdegutis** [comment]
This is the behavior I get, trying to scroll inside the box:

![ScreenRecording2019-12-06at12022](https://user-images.githubusercontent.com/4399718/70345028-a5543200-1820-11ea-9bcf-81ba7550d595.gif)

This is with [idea2.html](https://sdegutis.github.io/xterm-widget-test-case/idea2.html).

**Comment by Tyriar** [comment]
@sdegutis oh I see that now, maybe I had the first one opened. That's pretty weird behavior, don't you get the event in the decoration first and you can cancel it?

**Comment by sdegutis** [comment]
@Tyriar It doesn't get any scroll events. Tried by adding:

```js
    widget.addEventListener('scroll', (e) => {
      console.log(e);
    });
```

**Comment by Tyriar** [comment]
@sdegutis the viewport listens to scroll, shouldn't it's children get it first? 😕 

https://github.com/xtermjs/xterm.js/blob/bd0d267d972a1aab4777cddecf9bf8196d7aa44f/src/browser/Viewport.ts#L50

**Comment by sdegutis** [comment]
That's what I thought too. Was hoping there was some special thing you guys are doing and that I just didn't know about it. But I guess it's a mystery to all of us 😄 Looking into it now...

**Comment by sdegutis** [comment]
I don't see anything strange in Lifecycle.ts except that `useCapture` defaults to undefined, yet is still being passed. But that probably shouldn't matter... I think.

Also I may have found a bug: it checks if the handler is unset when disposing, but never unsets it.

**Comment by Tyriar** [comment]
We're planning on looking into this for Feb, there's 3 main ideas of what this encompasses so far:

- Buffer decorations (the main request)
- Gutter decorations (show things to the left of the buffer so it's not so cramped)
- Monaco-style scroll bar annotations for the decorations. We don't need to get as granular as monaco allows, but this is the basic idea
   ![image](https://user-images.githubusercontent.com/2193314/151197402-6356f009-e004-42f2-8a20-5f1779b41137.png)

**Comment by meganrogge** [comment]
We're thinking to support decorations in the buffer, gutter, and scrollbar (in that order). 

I have started working on the buffer decorations and am seeking feedback on this first pass at the API:

```TS
  /**
     * (EXPERIMENTAL) Adds a decoration to the terminal using
     *  @param decorationOptions, which takes a marker and an optional anchor, 
     *  width, height, and x offset from the anchor
     */
    registerDecoration(decorationOptions: IBufferDecorationOptions): IDecoration | undefined;
```


```TS
/**
   * Represents a disposable with an 
   * @param onDispose event listener and
   * @param isDisposed property.
   */
  export interface IDisposableWithEvent extends IDisposable {
    /**
     * Event listener to get notified when this gets disposed.
     */
    onDispose: IEvent<void>;

    /**
     * Whether this is disposed.
     */
    readonly isDisposed: boolean;
  }

  /**
   * Represents a decoration in the terminal that is associated with a particular marker and DOM element.
   */
  export interface IDecoration extends IDisposableWithEvent {
    /*
     * The marker for the decoration in the terminal.
     */
    readonly marker: IMarker;

    /**
     * An event fired when the decoration
     * is rendered, returns the dom element
     * associated with the decoration.
     */
    onRender: IEvent<HTMLElement>;

    /**
     * The HTMLElement that gets created after the 
     * first _onRender call, or undefined if accessed before
     * that.
     */
    element: HTMLElement | undefined;
  }

  export interface IBufferDecorationOptions {
    /**
     * The line in the terminal where
     * the decoration will be displayed
     */
    marker: IMarker;

    /*
     * Where the decoration will be anchored -
     * defaults to the left edge
     */
    anchor?: 'right' | 'left';

    /**
     * The width of the decoration, which defaults to 
     * cell width
     */
    width?: number;

    /**
     * The height of the decoration, which defaults to 
     * cell height
     */
    height?: number;

    /**
     * The x position offset relative to the anchor
     */ 
    x?: number;
  }
``` 

Here's what we're thinking for the gutter decorations:
```TS
    export interface IGutterDecorationOptions {
      /**
       * The line in the terminal where
       * the decoration will be displayed
       */
      startMarker: IMarker;
      /**
       * The end line in the terminal for
       * the decoration
       */
      endMarker: IMarker;
    }
```

**Comment by Tyriar** [comment]
> registerDecoration(decorationOptions: IBufferDecorationOptions): IDecoration | undefined;

Could callout when undefined happens in the jsdoc

> onDispose: IEvent<void>;
> onRender: IEvent<HTMLElement>;
> element: HTMLElement | undefined;

These should be readonly since xterm.js creates them.

> width?: number;
> height?: number;

Maybe stick to 1x1 now and consider this down the line? If we did have this I think specifying the width and height in cells would make the most sense.

**Comment by Eugeny** [comment]
Specifying the width (in cells, not pixels) is a must-have for keyword highlighting, prompt decoration and other applications.
